File: app/src/main/java/com/example/flicks/MainActivity.java
package com.example.flicks;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.example.flicks.adapter.RecyclerViewAdapter;
import com.example.flicks.api.ApiModule;
import com.example.flicks.api.ApiService;
import com.example.flicks.entity.NowPlaying;
import com.example.flicks.mapper.DataMapper;
import com.example.flicks.model.Movie;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity{

    protected  List<Movie> movieList;
    protected RecyclerViewAdapter adapter;
    protected DataMapper mapper;

    RecyclerView recyclerView;
    SwipeRefreshLayout swipeRefreshLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.rv);
        swipeRefreshLayout = findViewById(R.id.swipeRefresh);

        if(movieList == null){
            movieList = new ArrayList<>();
            setupList();
            getNowPlaying();
        }
        else adapter.setData(movieList);

    }
    private void getNowPlaying() {

        ApiModule.getInstance().getNowPlaying(ApiService.API_KEY)
                .enqueue(new Callback<NowPlaying>() {
                    @Override
                    public void onResponse(Call<NowPlaying> call, Response<NowPlaying> response) {
                        if(response.body() != null){
                            movieList = mapper.transform(response.body());
                            adapter.setData(movieList);
                        }
                    }

                    @Override
                    public void onFailure(Call<NowPlaying> call, Throwable t) {
                    }
                });

    }

    private void setupList() {
        mapper = new DataMapper();
        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(mLayoutManager);

        adapter = new RecyclerViewAdapter(MainActivity.this,movieList);
        adapter.setData(movieList==null ? new ArrayList<Movie>(): movieList);
        recyclerView.setAdapter(adapter);

        adapter.setListener(new RecyclerViewAdapter.IClickListener() {
            @Override
            public void onItemClick(Movie movie) {
                Bundle bundle = new Bundle();
                bundle.putParcelable("movie",movie);
                Intent detail =new Intent(MainActivity.this, DetailActivity.class);
                detail.putExtras(bundle);
                startActivity(detail);            }
        });

        this.recyclerView.setAdapter(adapter);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                adapter.clearData();
                getNowPlaying();
                swipeRefreshLayout.setRefreshing(false);

            }
        });

    }
}

File: app/src/main/java/com/example/flicks/MainActivity.kt
package com.example.flicks

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.flicks.adapter.RecyclerViewAdapter
import com.example.flicks.api.ApiModule
import com.example.flicks.api.ApiService
import com.example.flicks.entity.NowPlaying
import com.example.flicks.model.Movie
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class MainActivity : AppCompatActivity() {
    //
    private var movieList: MutableList<Movie>? = null
    private var adapter: RecyclerViewAdapter? = null
    private var recyclerView: RecyclerView? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        recyclerView = findViewById(R.id.rv)
        swipeRefreshLayout = findViewById(R.id.swipeRefresh)
        movieList = ArrayList()
        setupList()
        getNowPlaying()
    }

    private fun setupList() {
        // khai bao recycler view
        val mLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        recyclerView!!.layoutManager = mLayoutManager
        adapter = RecyclerViewAdapter(this@MainActivity, movieList
                ?: ArrayList()) { movie -> // truyền phim vào trong màn hình sau
            val bundle = Bundle()
            bundle.putParcelable("movie", movie)
            val detail = Intent(this@MainActivity, DetailActivity::class.java)
            detail.putExtras(bundle)
            startActivity(detail)
        }

        //set adapter cho recycler view
        recyclerView!!.adapter = adapter

        //
        swipeRefreshLayout!!.setOnRefreshListener {
            adapter!!.clearData()
            getNowPlaying()
            swipeRefreshLayout!!.isRefreshing = false
        }
    }

    // goi api để lấy danh sách phim
    private fun getNowPlaying() {

        // goi api để lấy danh sách phim
        ApiModule.instance?.getNowPlaying(ApiService.API_KEY)
                ?.enqueue(object : Callback<NowPlaying?> {
                    override fun onResponse(call: Call<NowPlaying?>, response: Response<NowPlaying?>) {
                        if (response.body() != null) {
                            adapter!!.setData(response.body()?.results as MutableList<Movie>)
                        }
                    }

                    override fun onFailure(call: Call<NowPlaying?>, t: Throwable) {
                        Log.d("Leo", "fail")
                    }
                })
    }
}